Updated financial aid review page frontend

alvinsiu opened the pull request:
#### What are the relevant tickets?
Closes #1045, closes #1214

#### What's this PR do?
Updates the financial aid review page frontend and backend:
- Moved `FinancialAidStatus` and `FinancialAidJustification` to `financialaid/constants.py`
- Added `FinancialAid.justification`
- Temporarily removed `FinancialAidStatus.REJECTED` by commenting out code as necessary as per request
- Modified `PUT` requests for `FinancialAidActionView` to `PATCH` requests to better follow REST, updating `FinancialAidActionSerializer`, its validation, and tests accordingly
- Modified the UI for the review page as requested
- Changed `FinancialAidMailView` to use a serializer and to take a `FinancialAid` object instead of email recipient for security and auto-message-generation reasons
- Linked backend functionality for emailing, approving, and marking documents as received to the frontend
- UI selectively shows/hides sections and functionality depending on the status selected to view
- Updated labels and copy as per request
- Added search by name functionality

#### Where should the reviewer start?

#### How should this be manually tested?

#### Any background context you want to provide?
Since everything was pretty integrated across different parts of the codebase, I didn't make separate issues for the smaller components for the sake of time, hence the larger PR.

Screenshots below in the comments.

#### Screenshots (if appropriate)

#### What GIF best describes this PR or how it makes you feel?
![image](https://cloud.githubusercontent.com/assets/1454606/18932988/a9f8fdba-85a0-11e6-856a-7f67c0ac609a.png)

alvinsiu commented:
@Ferdi @pdpinch Here's the updated frontend page as you requested.  I still need to hook in the backend, but I wanted to have this up for you to have a look at.

pdpinch commented:
- [ ] Remove "reject" button  /cc @roberthouse54 
- [ ] Rename "approve" checkmark with label "Save"  /cc @roberthouse54 
- [ ] Move "adjusted cost" column close to "Save" button
- [ ] Add a dropdown with a couple of predefined values that will allow reviewer to log a justification for adjusting the cost (e.g  "Docs not notarized" , "Insufficient Docs")

How hard will it be to update the predefined values for justification?

alvinsiu commented:
Pretty simple - just updating a constants file.

Ferdi commented:
@alvinsiu thanks !

Can we not add that top row for sorting and instead use the existing headers ? I understand it wont work for "location" but I dont think that's important.

alvinsiu commented:
Sure we can make that happen!

Ferdi commented:
What does "Search" box do ?   If it search on the name field, can we change the label to "Search by name" ?

alvinsiu commented:
The search box doesn't do anything at the moment - it was just on the mock, so I put it in there as a placeholder.  Otherwise, I don't think we've ever defined what that functionality should look like.

alvinsiu commented:
Updated screenshots:

![image](https://cloud.githubusercontent.com/assets/1454606/19096626/e0216560-8a6b-11e6-9d44-ed32af88eb85.png)
![image](https://cloud.githubusercontent.com/assets/1454606/19096632/e9a223c2-8a6b-11e6-96a6-48a620471f49.png)
![image](https://cloud.githubusercontent.com/assets/1454606/19096639/ff309746-8a6b-11e6-89cd-0f867ae99ffc.png)
![image](https://cloud.githubusercontent.com/assets/1454606/19096643/086e2706-8a6c-11e6-81f1-bb094382a726.png)
![image](https://cloud.githubusercontent.com/assets/1454606/19096655/172656a6-8a6c-11e6-92a9-a83e3556e342.png)
![image](https://cloud.githubusercontent.com/assets/1454606/19096658/1a7105b8-8a6c-11e6-83cf-317a4d03dce3.png)

alvinsiu commented:
Quite a few changes in this PR - a lot of them small tweaks uncovered while linking the back and frontends, and some of them slightly more involved.  Here are the general changes made:

- Moved `FinancialAidStatus` and `FinancialAidJustification` to `financialaid/constants.py`
- Added `FinancialAid.justification`
- Temporarily removed `FinancialAidStatus.REJECTED` by commenting out code as necessary as per request
- Modified `PUT` requests for `FinancialAidActionView` to `PATCH` requests to better follow REST, updating `FinancialAidActionSerializer`, its validation, and tests accordingly
- Modified the UI for the review page as requested
- Changed `FinancialAidMailView` to use a serializer and to take a `FinancialAid` object instead of email recipient for security and auto-message-generation reasons
- Linked backend functionality for emailing, approving, and marking documents as received to the frontend
- UI selectively shows/hides sections and functionality depending on the status selected to view
- Updated labels and copy as per request
- Added search by name functionality

noisecapella commented:
In the future please limit the amount of refactoring in these PRs unless it's related to the issue at hand. Sorry for being blunt but the financial aid APIs are a bit messy and I would rather do that kind of refactoring when we're not under major time pressure, so we can do it all at once and thoughtfully. It also increases the review workload when we're under a time crunch right now

noisecapella commented on financialaid/models.py:
```diff
@@ -138,7 +106,11 @@ class FinancialAid(TimestampedModel):
-        choices=[(status, status) for status in FinancialAidStatus.ALL_STATUSES],
+        choices=[
+            (status, status) for status in FinancialAidStatus.ALL_STATUSES
+            # The rejected status has been temporarily disabled as a valid status for FinancialAid
```
Please revert this change unless it's necessary for the work in this PR. We should not be putting in 'temporary' changes, especially right before a release. We should remove `rejected` with a separate PR

alvinsiu replied:
It's not a "temporary" change per se - it's supposed to be removed for 10/15, but I remember during a standup last week @pdpinch said to leave the code in for `rejected`, but commented out.

But you're right, we should remove `rejected` with a separate PR - will revert changes and do that there.

noisecapella commented on financialaid/serializers.py:
```diff
@@ -110,53 +113,70 @@ def save(self):
-        except TierProgram.DoesNotExist:
-            raise ValidationError("Financial Aid Tier does not exist for this program.")
+        # Required field
+        if data.get("action", None) is None:
```
`get` doesn't need to be passed `None`, it is the default

alvinsiu replied:
Will remove these instances

noisecapella commented on financialaid/templates/review_financial_aid.html:
```diff
@@ -1,116 +1,409 @@
+    
+    
+    {% if selected_status == statuses.PENDING_MANUAL_APPROVAL %}
+    /**
```
Can you move all this JS to a separate file in `static/js`? We also don't need to define the functions conditionally, it would be better if they were always defined

alvinsiu replied:
Will do!

noisecapella commented:
How does the user set justification for pending-manual-approval users?

![screenshot from 2016-10-05 11-21-13](https://cloud.githubusercontent.com/assets/863262/19119491/e43eb864-8aed-11e6-8127-a9d95d9e184b.png)

roberthouse54 commented:
@noisecapella I don't think we need a confirmation for either of these.

alvinsiu commented:
@noisecapella @roberthouse54 What if they accidentally click the wrong button for the wrong user?  Or they've accidentally tabbed over and hit enter?  It seems dangerous to me to not have a confirmation of some sort.

Thoughts?

roberthouse54 commented:
I guess I am ok with a confirmation dialog, but this will maybe get annoying to see every time. But you are probably right.

alvinsiu commented on financialaid/serializers.py:
```diff
@@ -110,53 +113,70 @@ def save(self):
+            if data.get("tier_program_id", None) is None:
+                raise ValidationError({"tier_program_id": "This field is required."})
+            if data.get("justification", None) is None:
+                raise ValidationError({"justification": "This field is required."})
```
These are used on the frontend.

The dicts vs strings is how DRF differentiates between field errors and non-field errors.

noisecapella replied:
Ah, I didn't know that

noisecapella commented:
When I search using the textbox I get this exception:

![screenshot from 2016-10-05 17-17-21](https://cloud.githubusercontent.com/assets/863262/19131981/9830c69c-8b1f-11e6-886c-acf4e34a2796.png)

alvinsiu commented:
Ah, forgot to remove django tags when moving the javascript.

noisecapella commented:
I'm getting an error when clicking 'mark docs received':

![screenshot from 2016-10-06 09-59-55](https://cloud.githubusercontent.com/assets/863262/19155258/b82e6972-8bab-11e6-9f5f-137c6bfdb40c.png)

It looks like there is django template markup in the JS file. This should be passed as a parameter instead

alvinsiu commented:
I should have caught that in testing, my bad - I've re-tested all of the actions, so it should work as expected now.

noisecapella commented on static/js/financial_aid_review.js:
```diff
@@ -0,0 +1,161 @@
+   * Marks documents as received for a financial aid application
+   * 
+   * @param financialAidId {number} Financial aid application id
+   * @param url {string} URL to submit request to
```
Can you add `action` here?

alvinsiu replied:
Yup!

noisecapella commented on static/js/financial_aid_review.js:
```diff
@@ -0,0 +1,161 @@
+   * Submits a financial aid application approval
+   * 
+   * @param financialAidId {number} Financial aid application id
+   * @param url {string} URL to submit request to
```
Here as well

alvinsiu replied:
Done!

noisecapella commented:
Can you download the jquery and bootstrap files that you're using and put them in `static/js/financial_aid`? Also please move the `financial_aid_review.js` there too

alvinsiu commented:
Sure - I'll do that once I finish running tests for merging with master!

noisecapella commented on financialaid/views.py:
```diff
@@ -164,43 +172,36 @@ def get_context_data(self, **kwargs):
         ).annotate(
```
We need to filter on `current=True` here

alvinsiu replied:
Good catch!

noisecapella commented on financialaid/templates/review_financial_aid.html:
```diff
@@ -1,116 +1,276 @@
               <td>
-                <button class="btn btn-default">
-                  <span class="glyphicon glyphicon-ok" style="color: #0074e1"></span>
+                {% if selected_status == statuses.PENDING_MANUAL_APPROVAL %}
```
Can you remove this if statement? It seems redundant

alvinsiu replied:
Ah, this was intentional, but I messed up fixing an earlier bug.  This is to show the justification for approved applications, but render the `<select>` if it's pending manual approval.  I'll make the change.

noisecapella commented on financialaid/views_test.py:
```diff
@@ -252,25 +256,52 @@ def test_review_financial_aid_view_with_invalid_filter_and_sorting(self):
+            Q(user__profile__first_name__icontains=search_query) | Q(user__profile__last_name__icontains=search_query),
+            tier_program__program_id=self.program.id,
+            status=FinancialAidStatus.AUTO_APPROVED
+        ).order_by("user__profile__first_name").values_list("id", flat=True)  # Default sort field
```
The default sort is `last_name`.

By the way, one of the problems with `UserFactory` is that it doesn't generate a profile using `ProfileFactory`, a blank profile is created automatically with a `post_save` signal. I think this works only because all profiles here are blank. Can you change instances of `FinancialAidFactory` to use profiles generated using `ProfileFactory`, maybe with a helper function?

alvinsiu replied:
Ah, good catch.

And yup, I can make changes to the FinancialAidFactory.

alvinsiu replied:
I made this change by overriding `FinancialAidFactory.create()` so that it retains the same workflow as every other factory.  It's basically the same thing as what you suggested, just not as a separate function to call.

Let me know if the comments I have on it aren't clear enough or if you think of any issues with this solution.

noisecapella replied:
Looks fine to me

noisecapella commented on financialaid/serializers.py:
```diff
@@ -119,37 +122,52 @@ class FinancialAidActionSerializer(serializers.Serializer):
-            raise ValidationError("Financial Aid Tier does not exist for this program.")
+        # Required field
+        if data.get("action") is None:
+            raise ValidationError({"action": "This field is required."})
```
Can you add tests to cover this line and the two other `This field is required` validation errors below?

noisecapella commented on financialaid/views_test.py:
```diff
@@ -204,13 +208,13 @@ def test_review_financial_aid_view_with_filter_and_sorting(self):
         FinancialAidFactory.create(tier_program=self.tier_programs["0k"], status=FinancialAidStatus.AUTO_APPROVED)
```
Can we add more `FinancialAid`s with the same status? With one element for each we can't verify sort

alvinsiu replied:
Sounds good!

noisecapella commented on financialaid/views_test.py:
```diff
@@ -252,25 +256,52 @@ def test_review_financial_aid_view_with_invalid_filter_and_sorting(self):
+        """
+        Tests that ReviewFinancialAidView returns the expected results with search
+        """
+        FinancialAidFactory.create(tier_program=self.tier_programs["0k"], status=FinancialAidStatus.AUTO_APPROVED)
```
Can you add more copies here too?

noisecapella commented:
It looks like there's a missing migration due to the status order change

noisecapella commented:
Can you rebase please?

alvinsiu commented on static/js/financial_aid/bootstrap.min.js:
```diff
@@ -0,0 +1,8 @@
+/* eslint-disable */
```
Yup!

alvinsiu commented:
Wow, just rebased an hour ago and already conflicts :/

noisecapella commented:
Ok I think that's everything :+1:

alvinsiu commented:
Thanks for the really thorough review on this really long PR!
